f600233d1b154d507b782f84604bcb12/update/sys_hub_action_type_definition_7232d6c92fe41110e58359a72799b6f0.xml

```diff
@@ -0,0 +1,1331 @@
+        <sys_id>dd88d6412f281110e58359a72799b6fa</sys_id>
+        <sys_mod_count>3</sys_mod_count>
+        <sys_updated_by>jeff.segal-ctr@ccpo.mil</sys_updated_by>
+        <sys_updated_on>2022-07-06 02:09:15</sys_updated_on>
+        <value>function addCostRecord(result, csp, organization, portfolio, taskOrder, clin, isActual) {
+  var gdt = new GlideDateTime();
+  gdt.setDisplayValue(result.month, "yyyy-MM");
+  
+  // Check to see if we already have a record for this Task Order / CLIN / Year-Month Tuple
+  var cost = new GlideRecord('x_g_dis_atat_costs');
+  cost.addQuery("task_order_number", taskOrder.task_order_number);
+  cost.addQuery("clin", clin.clin_number);
+  cost.addQuery("year_month", gdt.getDate());
+  cost.query();
+  
+  var isUpdate = false;
+  if (cost.hasNext()) {
+    cost.next();
+    isUpdate = true;       
+  }
+  else {
+    cost.newRecord();
+    isUpdate = false;       
+  }
+
+  cost.clin = clin.clin_number;
+  cost.csp = csp.sys_id;
+  cost.is_actual = isActual;
+  cost.organization = organization.sys_id;
+  cost.portfolio = portfolio.sys_id;
+  cost.service_agency = organization.service_agency;
+  cost.task_order_number = taskOrder.task_order_number;
+  cost.value = result.value;
+  cost.year_month = gdt;
+  
+  if (isUpdate) {
+    cost.update();
+  }
+  else {
+    cost.insert();
+  }
+  
+}
+
+function handleCostObject(costObject) {
+  var csp = new GlideRecord('x_g_dis_atat_cloud_service_provider');
+  csp.get("name", costObject.content.request.targetCsp.name);
+
+  var portfolio = new GlideRecord('x_g_dis_atat_portfolio');
+  portfolio.get("csp_portfolio_id", costObject.content.request.portfolioId);
+  
+  var organization = portfolio.acquisition_package.organization;
+    
+  var taskOrders = costObject.content.response.task_orders;
+  for (var i = 0; i &lt; taskOrders.length; i++) {
+    var taskOrder = taskOrders[i];
+    var clins = taskOrder.clins;
+    for (var j = 0; j &lt; clins.length; j++) {
+      var clin = clins[j];
+      
+      var actuals = clin.actual;
+      for (var k = 0; k &lt; actuals.length; k++) {
+        var actual = actuals[k];
+        var actual_results = actual.results;
+        for (var l = 0; l &lt; actual_results.length; l++) {
+          addCostRecord(actual_results[l], csp, organization, portfolio, taskOrder, clin, true);
+        }      
+      }
+      
+      var forecasts = clin.forecast;
+      for (var k = 0; k &lt; forecasts.length; k++) {
+        var forecast = forecasts[k];
+        var forecast_results = forecast.results;
+        for (var l = 0; l &lt; forecast_results.length; l++) {
+          addCostRecord(forecast_results[l], csp, organization, portfolio, taskOrder, clin, false);
+        }        
+      }      
+    }
+  }  
+}
+
+(function execute(inputs, outputs) {
+  var costObjects = JSON.parse(inputs.cost_response);
+  for (var i = 0; i &lt; costObjects.length; i++) {
+	handleCostObject(costObjects[i]);
+  }
```

Comment:
Just a suggestion, but I believe you can use a `.map` here to help simplify the code. Probably also in the `handleCostObject` function to reduce some of the nesting of the `for` loops making it a little more readable (might require breaking down into smaller functions though if you don't want to do `function() {}` inside the map).
```suggestion
    costObjects.map(handleCostObject)
```

Reply:
I think I got annoyed trying to use `forEach` earlier and gave up, but when I returned to it tonight I realized that it was just that arrow functions weren't supported. So here's what I ended up with:

```js
  taskOrders.forEach(function (taskOrder) {
    taskOrder.clins.forEach(function (clin) {
      clin.actual.forEach(function (actual) {
        actual.results.forEach(function (result) {
          addCostRecord(result, csp, organization, portfolio, taskOrder, clin, true);
        })
      })
      clin.forecast.forEach(function (forecast) {
        forecast.results.forEach(function (result) {
          addCostRecord(result, csp, organization, portfolio, taskOrder, clin, false);
        })
      })
    })
  });
```